File: cpp/examples/task_board/task_dialog/task_dialog.cpp
#include "task_dialog.hpp"

#include "ui/elements/ComboBox.hpp"

#include <algorithm>
#include <cctype>
#include <cstddef>
#include <string_view>
#include <utility>

namespace task_board_example
{

namespace
{

constexpr std::string_view kDefaultStatus = "Todo";
constexpr std::string_view kDefaultAssignee = "Ava";
constexpr std::string_view kDefaultPriority = "Med";
constexpr std::string_view kDoneStatus = "Done";
constexpr std::string_view kNewTaskTitle = "New task";
constexpr std::string_view kEditTaskTitle = "Edit task";
constexpr std::string_view kTitleRequiredMessage = "Title is required.";

[[nodiscard]] std::string trim(std::string_view text)
{
    std::size_t begin = 0;
    while (begin < text.size() && std::isspace(static_cast<unsigned char>(text[begin])))
    {
        ++begin;
    }

    std::size_t end = text.size();
    while (end > begin && std::isspace(static_cast<unsigned char>(text[end - 1])))
    {
        --end;
    }

    return std::string{text.substr(begin, end - begin)};
}

void selectComboValue(ui::ComboBox& comboBox, std::string_view value)
{
    const auto& options = comboBox.options();
    for (std::size_t index = 0; index < options.size(); ++index)
    {
        if (options[index] == value)
        {
            comboBox.setSelectedIndex(static_cast<int>(index));
            return;
        }
    }
    comboBox.setSelectedIndex(0);
}

[[nodiscard]] std::string selectedComboValue(const ui::ComboBox& comboBox, std::string_view fallback)
{
    const auto& options = comboBox.options();
    const int index = comboBox.selectedIndex();
    if (index < 0 || index >= static_cast<int>(options.size()))
    {
        return std::string{fallback};
    }
    return options[static_cast<std::size_t>(index)];
}

}  // namespace

TaskDialog::TaskDialog(std::optional<TaskDialogResult> task, TaskDialogCallback onClose)
    : initialTask_(std::move(task)),
      onClose_(std::move(onClose))
{
    if (initialTask_.has_value())
    {
        loadTask(*initialTask_);
    }
    else
    {
        header->setText(std::string{kNewTaskTitle});
        selectComboValue(*project_status, kDefaultStatus);
        selectComboValue(*owner, kDefaultAssignee);
        selectComboValue(*priority, kDefaultPriority);
    }
}

bool TaskDialog::submit()
{
    const std::string error = validationError();
    if (!error.empty())
    {
        message->setText(error);
        return false;
    }

    close(formResult());
    return true;
}

void TaskDialog::cancel()
{
    close(std::nullopt);
}

void TaskDialog::onAddBtnClick()
{
    (void)submit();
}

void TaskDialog::onCancelBtnClick()
{
    cancel();
}

void TaskDialog::loadTask(const TaskDialogResult& task)
{
    header->setText(std::string{kEditTaskTitle});
    name->setValue(task.title);
    selectComboValue(*project_status, task.status.empty() ? kDefaultStatus : std::string_view(task.status));
    selectComboValue(*owner, task.assignee.empty() ? kDefaultAssignee : std::string_view(task.assignee));
    selectComboValue(*priority, task.priority.empty() ? kDefaultPriority : std::string_view(task.priority));
    note->setValue(task.description);
}

void TaskDialog::close(std::optional<TaskDialogResult> result)
{
    result_ = std::move(result);
    closed_ = true;
    if (onClose_)
    {
        onClose_(result_);
    }
}

TaskDialogResult TaskDialog::formResult() const
{
    const std::string statusValue = selectedComboValue(*project_status, kDefaultStatus);
    return TaskDialogResult{
        .id = initialTask_.has_value() ? initialTask_->id : std::string{},
        .title = trim(name->value()),
        .status = statusValue,
        .assignee = selectedComboValue(*owner, kDefaultAssignee),
        .priority = selectedComboValue(*priority, kDefaultPriority),
        .description = trim(note->value()),
        .done = statusValue == kDoneStatus,
    };
}

std::string TaskDialog::validationError() const
{
    if (trim(name->value()).empty())
    {
        return std::string{kTitleRequiredMessage};
    }
    return {};
}

}  // namespace task_board_example

File: go/examples/task_board/task_dialog/task_dialog.go
package task_dialog

import (
	"strings"

	uimd "uimd"
)

const (
	defaultStatus        = "Todo"
	defaultAssignee      = "Ava"
	defaultPriority      = "Med"
	doneStatus           = "Done"
	newTaskTitle         = "New task"
	editTaskTitle        = "Edit task"
	titleRequiredMessage = "Title is required."
)

type TaskDialogResult struct {
	ID          string
	Title       string
	Status      string
	Assignee    string
	Priority    string
	Description string
	Done        bool
}

type TaskDialog struct {
	*TaskDialogUI
	initial *TaskDialogResult
	result  *TaskDialogResult
	closed  bool
}

func NewTaskDialog(task *TaskDialogResult) *TaskDialog {
	dialog := &TaskDialog{TaskDialogUI: NewTaskDialogUI()}
	dialog.SetEventHandler(dialog)
	if task != nil {
		copyTask := *task
		dialog.initial = &copyTask
		dialog.LoadTask(copyTask)
	} else {
		dialog.Header.SetText(newTaskTitle)
		selectComboValue(dialog.ProjectStatus, defaultStatus)
		selectComboValue(dialog.Owner, defaultAssignee)
		selectComboValue(dialog.Priority, defaultPriority)
	}
	return dialog
}

func (dialog *TaskDialog) OnAddBtnClick() {
	dialog.Submit()
}

func (dialog *TaskDialog) OnCancelBtnClick() {
	dialog.Cancel()
}

func (dialog *TaskDialog) Submit() bool {
	if message := dialog.validationError(); message != "" {
		dialog.Message.SetText(message)
		return false
	}
	result := dialog.formResult()
	dialog.result = &result
	dialog.closed = true
	return true
}

func (dialog *TaskDialog) Cancel() {
	dialog.result = nil
	dialog.closed = true
}

func (dialog *TaskDialog) Result() *TaskDialogResult {
	if dialog.result == nil {
		return nil
	}
	result := *dialog.result
	return &result
}

func (dialog *TaskDialog) LoadTask(task TaskDialogResult) {
	dialog.Header.SetText(editTaskTitle)
	dialog.Name.SetValue(task.Title)
	if task.Status == "" {
		task.Status = defaultStatus
	}
	if task.Assignee == "" {
		task.Assignee = defaultAssignee
	}
	if task.Priority == "" {
		task.Priority = defaultPriority
	}
	selectComboValue(dialog.ProjectStatus, task.Status)
	selectComboValue(dialog.Owner, task.Assignee)
	selectComboValue(dialog.Priority, task.Priority)
	dialog.Note.SetValue(task.Description)
}

func (dialog *TaskDialog) formResult() TaskDialogResult {
	status := selectedComboValue(dialog.ProjectStatus, defaultStatus)
	result := TaskDialogResult{
		Title:       strings.TrimSpace(dialog.Name.Value),
		Status:      status,
		Assignee:    selectedComboValue(dialog.Owner, defaultAssignee),
		Priority:    selectedComboValue(dialog.Priority, defaultPriority),
		Description: strings.TrimSpace(dialog.Note.Value),
		Done:        status == doneStatus,
	}
	if dialog.initial != nil {
		result.ID = dialog.initial.ID
	}
	return result
}

func (dialog *TaskDialog) validationError() string {
	if strings.TrimSpace(dialog.Name.Value) == "" {
		return titleRequiredMessage
	}
	return ""
}

func selectComboValue(combo *uimd.ComboBox, value string) {
	for index, option := range combo.Options {
		if option == value {
			combo.SetSelectedIndex(index)
			return
		}
	}
	combo.SetSelectedIndex(0)
}

func selectedComboValue(combo *uimd.ComboBox, fallback string) string {
	index := combo.SelectedIndex
	if index < 0 || index >= len(combo.Options) {
		return fallback
	}
	return combo.Options[index]
}
